### PR TITLE
Improve Error Handling for Tags

### PR DESCRIPTION
**Pull request by jycor**
Changes: 
- Prevent cloning from a tag through `--branch` option
- Return detached head error for checking out branch through cli

Fixes: https://github.com/dolthub/dolt/issues/8377

### CODE REVIEW

**Comment by coffeegoddd** [comment]
@jycor __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>8ed8710</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>8ed8710</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@jycor __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>cde716f</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>cde716f</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@jycor __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>685c7a5</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>685c7a5</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/libraries/doltcore/sqle/dprocedures/dolt_checkout.go [hunk 1]
```diff
@@ -290,11 +290,15 @@ func checkoutRemoteBranch(ctx *sql.Context, dSess *dsess.DoltSession, dbName str
 	}
 
 	if len(remoteRefs) == 0 {
+		if isTag, err := actions.IsTag(ctx, dbData.Ddb, branchName); err != nil {
+			return "", err
+		} else if isTag && apr.Contains(cli.MoveFlag) {
+			return "", fmt.Errorf(`dolt does not support a detached head state. To checkout a tag, run: 
+	dolt checkout %s -b {new_branch_namee}`, branchName)
```
Comment (fulghum) [inline go/libraries/doltcore/sqle/dprocedures/dolt_checkout.go:297]:
```suggestion
	dolt checkout %s -b {new_branch_name}`, branchName)
```

File: go/libraries/doltcore/sqle/dprocedures/dolt_checkout.go [hunk 2]
```diff
@@ -290,11 +290,15 @@ func checkoutRemoteBranch(ctx *sql.Context, dSess *dsess.DoltSession, dbName str
 	}
 
 	if len(remoteRefs) == 0 {
+		if isTag, err := actions.IsTag(ctx, dbData.Ddb, branchName); err != nil {
+			return "", err
+		} else if isTag && apr.Contains(cli.MoveFlag) {
```
Comment (fulghum) [inline go/libraries/doltcore/sqle/dprocedures/dolt_checkout.go:295]: Do we need to only execute this logic to when `cli.MoveFlag` is present? My understanding is that we don't document the `MoveFlag`, but the CLI uses it when it calls `dolt_checkout()` to move the working set contents? It seems like we'd want to return the same error message if you tried to check out a tag from the SQL interface, too, right? I may just be missing something on how the `MoveFlag` is used though.

File: go/libraries/doltcore/sqle/dprocedures/dolt_checkout.go (hunk 1 above)
Comment (fulghum) [reply go/libraries/doltcore/sqle/dprocedures/dolt_checkout.go:297]: Nice error message btw! Really, **really** helpful to tell users exactly what command they need to run whenever we can.

File: go/libraries/doltcore/sqle/enginetest/dolt_queries.go [hunk 3]
```diff
@@ -4678,6 +4678,36 @@ var DoltTagTestScripts = []queries.ScriptTest{
...
+				Query: "select * from test;",
+				Expected: []sql.Row{
+					{0},
+					{1},
+					{2},
+				},
+			},
+			{
+				Query:          "call dolt_checkout('v1');",
+				ExpectedErrStr: "error: could not find v1",
```
Comment (fulghum) [inline go/libraries/doltcore/sqle/enginetest/dolt_queries.go:4707]: (minor) Might be nice if there was a more prescriptive error message here, like we use in `dolt checkout` to tell the user exactly what they need to run in order to checkout a tag as a new branch.

**Review by fulghum — APPROVED** [review]
Looks good!

**Comment by coffeegoddd** [comment]
@jycor __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>1b36adf</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>1b36adf</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>d309e8d</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>d309e8d</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Pull request merged**